lib/properties.ex

```diff
@@ -186,52 +186,92 @@ defmodule PropCheck.Properties do
-            "Counter example stored."
-          else
-            "Counter example NOT stored, :store_counter_example is set to false."
-          end
+  # In this method a new found counter example is added to `CounterStrike`. Note that
+  # some macros such as exists/2 do not return counter examples when they fail.
+  defp handle_check_results(true, _name, _opts, _should_fail = false, _store_counter_example?) do
+    true
+  end
+
+  defp handle_check_results(true, name, _opts, _should_fail = true, _store_counter_example?) do
+    raise ExUnit.AssertionError,
+      message: "Property #{mfa_to_string(name)} should fail, but succeeded for all test data :-(",
+      expr: nil
+  end
 
-        raise ExUnit.AssertionError, [
-          message: """
-          Property #{mfa_to_string name} failed. Counter-Example is:
-          #{inspect counter_example, pretty: true}
+  defp handle_check_results(
+         error = {:error, _},
+         name,
+         _opts,
+         _should_fail,
+         _store_counter_example?
+       ) do
+    raise ExUnit.AssertionError,
+      message: "Property #{mfa_to_string(name)} failed with an error: #{inspect(error)}",
+      expr: nil
+  end
+
+  defp handle_check_results(
+         counter_example,
+         _name,
+         _opts,
+         _should_fail = true,
+         _store_counter_example?
+       )
+       when is_list(counter_example) do
+    true
+  end
 
-          #{counter_example_message}
-          """ |> add_additional_output(opts),
-          expr: nil]
-      {:rerun_failed, counter_example} when is_list(counter_example) ->
+  defp handle_check_results(counter_example, name, opts, _should_fail, store_counter_example?)
+       when is_list(counter_example) do
+    counter_example_message =
+      if store_counter_example? do
         CounterStrike.add_counter_example(name, counter_example)
-        raise ExUnit.AssertionError, [
-          message: """
-          Property #{mfa_to_string name} failed. Counter-Example is:
-          #{inspect counter_example, pretty: true}
-
-          Consider running `MIX_ENV=test mix propcheck.clean` if a bug in a generator was
-          identified and fixed. PropCheck cannot identify changes to generators. See
-          https://github.com/alfert/propcheck/issues/30 for more details.
-          """ |> add_additional_output(opts),
-          expr: nil]
-    end
+        "Counter example stored."
+      else
+        "Counter example NOT stored, :store_counter_example is set to false."
+      end
+
+    raise ExUnit.AssertionError,
+      message:
+        """
+        Property #{mfa_to_string(name)} failed. Counter-Example is:
+        #{inspect(counter_example, pretty: true)}
+
+        #{counter_example_message}
+        """
+        |> add_additional_output(opts),
+      expr: nil
+  end
+
+  defp handle_check_results(
+         {:rerun_failed, counter_example},
+         name,
+         opts,
+         _should_fail,
+         _store_counter_example?
```

Comment:
Maybe should_fail & store_counter_example? should be moved to a map? In most clauses they are not used. Would make code clearer.

Reply:
Good idea. I was wondering how to make this a little bit prettier.

Reply:
Maybe it would even make sense to put all arguments into a map? We can still match on those parts which are needed? @alfert What do you think?

Reply:
I went ahead and turned all arguments into a map with ecc7a08. What do you think, is this better?

Reply:
I would extract `result` as a separate argument and keep the pipe flow.

Reply:
Introduced the pipe again with fe8ed8e.

Reply:
Looks nice.

Reply:
The pipeline failed again in targeted pbt. :man_shrugging:

Reply:
It's a flaky test, no worries.  I have fix for that in a PR I'll submit later.

Reply:
Oh cool, I am very interested into this "take this, flaky test" PR!

Good work, by the way!

Reply:
@alfert don't be ;) It's an _"engineering"_ solution, not a _"scientific"_ one.